### PR TITLE
Allow IAM username suffix to be added to S3 scratch URL

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This PR retires the scratch-s3-bucket script and introduces the ability to use an IAM username as part of the scratch bucket URL with appropriate config files in place.  I'll work to get those set up so we can release this and have it work for our internal use.

I also moved the scratch bucket calculation over to BaseCreds, and renamed the functions to include the word 'default'.  Your AWS IAM username depends on which AWS creds you are using.  We always use the default AWS creds when loading/unloading on Redshift.  That might not be true in the future, and once we provide more flexibility there, we'll want to extend this code to take in the appropriate credentials name for the AWS creds.

### CODE REVIEW

File: records_mover/session.py [hunk 1]
```diff
@@ -156,6 +163,9 @@ def __init__(self,
                  default_aws_creds_name: Union[None, str, PleaseInfer] = PleaseInfer.token,
                  default_gcp_creds_name: Union[None, str, PleaseInfer] = PleaseInfer.token,
                  session_type: Union[str, PleaseInfer] = PleaseInfer.token,
+                 # TODO: FIgure out if this should be
```
Comment (bluelabsbutler) [inline records_mover/session.py:166]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 2]
```diff
@@ -177,6 +187,10 @@ def __init__(self,
         if default_gcp_creds_name is PleaseInfer.token:
             default_gcp_creds_name = _infer_default_gcp_creds_name(session_type)
 
+        # TODO: Move this logic to creds interface
```
Comment (bluelabsbutler) [inline records_mover/session.py:190]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 3]
```diff
@@ -185,11 +199,10 @@ def __init__(self,
...
                                  default_boto3_session=default_boto3_session,
                                  default_gcp_creds=default_gcp_creds,
-                                 default_gcs_client=default_gcs_client)
-
-        if scratch_s3_url is PleaseInfer.token:
-            scratch_s3_url = _infer_scratch_s3_url(session_type)
+                                 default_gcs_client=default_gcs_client,
+                                 scratch_s3_url=scratch_s3_url)
 
+        # TODO: get this from creds
```
Comment (bluelabsbutler) [inline records_mover/session.py:205]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/creds/base_creds.py [hunk 4]
```diff
@@ -1,7 +1,10 @@
 from db_facts.db_facts_types import DBFacts
 from .database import db_facts_from_env
 from typing import TYPE_CHECKING, Iterable, Union, Optional
-from records_mover.mover_types import NotYetFetched
+from records_mover.mover_types import PleaseInfer
+from config_resolver import get_config
+import os
+import subprocess
```
Comment (bluelabsbutler) [inline records_mover/creds/base_creds.py:7]: F401 'subprocess' imported but unused

File: records_mover/creds/base_creds.py [hunk 5]
```diff
@@ -140,11 +147,49 @@ def default_gcs_client(self) -> Optional['google.cloud.storage.Client']:
...
+                if append_iam_user_id_str is not None:
+                    if append_iam_user_id_str == 'true':
+                        append_iam_user_id = True
+                    elif append_iam_user_id_str == 'false':
+                        append_iam_user_id = False
+                    else:
+                        raise SyntaxError("Could not understand append_iam_user value - "
+                                          "Please set to true or false")
+                    if append_iam_user_id:
+                        # TODO: Make sure boto3_session is not None
```
Comment (bluelabsbutler) [inline records_mover/creds/base_creds.py:180]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/creds/base_creds.py [hunk 6]
```diff
@@ -140,11 +147,48 @@ def default_gcs_client(self) -> Optional['google.cloud.storage.Client']:
...
+            s3_scratch_url: Optional[str] = aws_cfg.get('s3_scratch_url')
+            if s3_scratch_url is None:
+                s3_scratch_url_prefix: Optional[str] =\
+                    aws_cfg.get('s3_scratch_url_appended_with_iam_user_id')
+                if s3_scratch_url_prefix is not None:
+                    if boto3_session is None:
+                        logger.warning('Cannot generate S3 scratch URL with IAM user ID, '
+                                       'as I have no IAM user ID')
+                        return None
+                    # TODO: Make sure boto3_session is not None
```
Comment (bluelabsbutler) [inline records_mover/creds/base_creds.py:177]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/creds/base_creds.py [hunk 7]
```diff
@@ -140,11 +147,48 @@ def default_gcs_client(self) -> Optional['google.cloud.storage.Client']:
...
+                    sts_client = boto3_session.client('sts')
+                    caller_identity = sts_client.get_caller_identity()
+                    arn = caller_identity['Arn']
+                    # TODO: What happens if nothing follows this?
```
Comment (bluelabsbutler) [inline records_mover/creds/base_creds.py:181]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: metrics/bigfiles_high_water_mark [hunk 8]
```diff
@@ -1 +1 @@
-981
+982
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: I added a comment to setup.py to explain a new version constraint.

File: records_mover/creds/base_creds.py [hunk 9]
```diff
@@ -1,7 +1,9 @@
 from db_facts.db_facts_types import DBFacts
 from .database import db_facts_from_env
 from typing import TYPE_CHECKING, Iterable, Union, Optional
-from records_mover.mover_types import NotYetFetched
+from records_mover.mover_types import PleaseInfer
```
Comment (vinceatbluelabs) [inline records_mover/creds/base_creds.py:4]: I standardized on the latter here; the difference in semantics was so subtle as to be meaningless for the uses here.

**Pull request merged**